lib/liquid_voting/delegations.ex

```diff
@@ -185,19 +186,46 @@ defmodule LiquidVoting.Delegations do
   def upsert_delegation(%{delegator_id: delegator_id, delegate_id: delegate_id} = attrs) do
     proposal_url = Map.get(attrs, :proposal_url)
 
-    Delegation
-    |> where(delegator_id: ^delegator_id)
-    |> Repo.all()
-    |> resolve_conflicts(delegate_id, proposal_url)
-    |> case do
-      {:ok, delegations} ->
-        delegations
-        |> find_similar_delegation_or_return_new_struct(proposal_url)
-        |> Delegation.changeset(attrs)
-        |> Repo.insert_or_update()
-
-      {:error, %{message: message, details: details}} ->
-        {:error, %{message: message, details: details}}
+    with {:ok} <- check_vote_conflict(delegator_id, proposal_url) do
```

Comment:
[`with`](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1) doesn't require an explicit else block for errors, unless you want to specify a particular error.  

That's why I chose to use `this` - it enables the passing on of an error from check_vote_conflict/2, without creating a nested conditional structure.

The error is demonstrated in the 2 tests in this PR adds which assert this error is thrown, and also in manual testing.

Reply:
I know `with` doesn't require an `else` block :) But I'd still like to understand where the error tuple is handled. It should be explicitly visible

Btw, "thrown" is said of exceptions, not returns

Reply:
You want an explicit else block in `upsert_delegation/1`?  Like this?
```
with {:ok} <- check_vote_conflict(delegator_id, proposal_url) do
      Delegation
      ...
      |> case do
        {:ok, delegations} ->
          ...

        {:error, %{message: message, details: details}} ->
          {:error, %{message: message, details: details}}
      end
else 
  {:error, message: message, details: details} ->
        {:error, message: message, details: details}
end
``

Reply:
No, I was just trying to understand code execution flow. It wasn't immediately obvious (to me) where the error went. Might just a be another symptom the error handling needs a refactor.

I've now fetched the PR branch locally and I'm having a look in my editor. What I'm understanding is `upsert_delegation/1` will return the error on line 141:

```elixir
|> Multi.run(:upsert_delegation, fn _repo, changes ->
  delegation_attrs
  ...
  |> upsert_delegation()
```

And get matched in line 148:

```elixir
{:error, :upsert_delegation, value, _} ->
  {:error, value}
```

I'm going to close this thread now and merge the PR, we can continue the code execution visibility discussion on refactor and error format issues.

Reply:
Yes, I think it will be much simpler when I refactor all the different error formats to one common one, as then we can simply have 2, chained, `with` statements here, with a single `else` statement for the common error format. :+1: